cloudinit/sources/DataSourceAzure.py

```diff
@@ -660,10 +700,292 @@ def setup(self, is_new_instance):
+                # won't be in primary_nic_found = false state for long.
+                if not primary_nic_found:
+                    LOG.info("Checking if %s is the primary nic",
+                             ifname)
+                    (primary_nic_found, expected_nic_count) = (
+                        self._check_if_nic_is_primary(ifname))
+
+                # Exit criteria: check if we've discovered all nics
+                if (expected_nic_count != -1
+                        and len(nics_found) >= expected_nic_count):
+                    LOG.info("Found all the nics for this VM.")
+                    break
+
+        except AssertionError as error:
+            report_diagnostic_event(error, logger_func=LOG.error)
+
+    @azure_ds_telemetry_reporter
+    def _wait_for_all_nics_ready(self):
+        """Wait for nic(s) to be hot-attached. There may be multiple nics
+           depending on the customer request.
+           But only primary nic would be able to communicate with wireserver
+           and IMDS. So we detect and save the primary nic to be used later.
+        """
+
+        try:
+            nl_sock = netlink.create_bound_netlink_socket()
+
+            report_ready_marker_present = bool(
+                os.path.isfile(REPORTED_READY_MARKER_FILE))
+
+            # Report ready if the marker file is not already present.
+            # The nic of the preprovisioned vm gets hot-detached as soon as
+            # we report ready. So no need to save the dhcp context.
+            self._report_ready_if_needed()
+
+            has_nic_been_detached = bool(
+                os.path.isfile(REPROVISION_NIC_DETACHED_MARKER_FILE))
+
+            if not has_nic_been_detached:
+                LOG.info("NIC has not been detached yet.")
+                self._wait_for_nic_detach(nl_sock)
+
+            # If we know that the preprovisioned nic has been detached, and we
+            # still have a fallback nic, then it means the VM must have
+            # rebooted as part of customer assignment, and all the nics have
+            # already been attached by the Azure platform. So there is no need
+            # to wait for nics to be hot-attached.
+            if not self.fallback_interface:
+                self._wait_for_hot_attached_nics(nl_sock)
+            else:
+                report_diagnostic_event("Skipping waiting for nic attach "
+                                        "because we already have a fallback "
+                                        "interface. Report Ready marker "
+                                        "present before detaching nics: %s" %
+                                        report_ready_marker_present,
+                                        logger_func=LOG.info)
+        except netlink.NetlinkCreateSocketError as e:
+            report_diagnostic_event(e, logger_func=LOG.warning)
+            raise
+        finally:
+            if nl_sock:
```

Comment:
If we reach this line because `create_bound_netlink_socket` raises an exception, we'll get a `NameError: name 'nl_sock' is not defined`:

```py
In [1]: def create_bound_netlink_socket(): 
   ...:     raise Exception("oh no") 
   ...:                                                                                                                                                                                                                                                                                   

In [2]: try: 
   ...:     nl_sock = create_bound_netlink_socket() 
   ...: except Exception as exc: 
   ...:     print("oh no") 
   ...:     raise 
   ...: finally: 
   ...:     if nl_sock: 
   ...:         print("nl_sock.close()") 
   ...:                                                                                                                                                                                                                                                                                   
oh no
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-2-8878d31775d2> in <module>
      1 try:
----> 2     nl_sock = create_bound_netlink_socket()
      3 except Exception as exc:

<ipython-input-1-46a18ca13bbb> in create_bound_netlink_socket()
      1 def create_bound_netlink_socket():
----> 2     raise Exception("oh no")
      3 

Exception: oh no

During handling of the above exception, another exception occurred:

NameError                                 Traceback (most recent call last)
<ipython-input-2-8878d31775d2> in <module>
      5     raise
      6 finally:
----> 7     if nl_sock:
      8         print("nl_sock.close()")
      9 

NameError: name 'nl_sock' is not defined
```

This suggests we're missing a test which would have triggered this exception, so I'd like to see a fix (should be simple!) and a test added.

Reply:
Thanks for catching this. Initialized nl_sock to None before entering the try block. Added a test that validates this.